contracts/actions/mcd/McdBoostComposite.sol

```diff
@@ -0,0 +1,213 @@
+            _subData,
+            _returnValues
+        );
+
+        (bytes memory logData, uint256 suppliedAmount) = _boost(boostParams);
+        emit ActionEvent("McdBoostComposite", logData);
+
+        return bytes32(suppliedAmount);
+    }
+
+    /// @inheritdoc ActionBase
+    function executeActionDirect(bytes memory _callData)
+        public
+        payable
+        virtual
+        override(ActionBase, DFSSell, GasFeeTaker)
+    {
+        BoostParams memory boostParams = _parseCompositeParams(_callData);
+        (bytes memory logData, ) = _boost(boostParams);
+        logger.logActionDirectEvent("McdBoostComposite", logData);
+    }
+
+    /// @notice Executes boost logic
+    function _boost(BoostParams memory _boostParams)
+        internal
+        returns (bytes memory logData, uint256 supplyAmount)
+    {
+        assert(_boostParams.exchangeData.srcAddr == DAI_ADDR);
+        (address urn, bytes32 ilk) = getUrnAndIlk(MCD_MANAGER_ADDR, _boostParams.vaultId);
+
+        uint256 collateralBefore;
+        if (_boostParams.gasUsed != 0) {
+            (collateralBefore, ) = getCdpInfo(
+                    IManager(MCD_MANAGER_ADDR),
+                    _boostParams.vaultId,
+                    ilk
+                );
+        }
+
+        address collateralAsset = _boostParams.exchangeData.destAddr;
+        uint256 boostAmount = _boostParams.exchangeData.srcAmount;
+
+        if (_boostParams.flAddr == address(0)) {
+            _drawDebt(_boostParams, urn, ilk);
+        }
+
+        // Sell debt asset for collateral asset
+        (uint256 exchangedAmount, ) = _dfsSell(
+            _boostParams.exchangeData,
+            address(this),
+            address(this),
+            false
+        );
+
+        // Take gas fee if part of strategy
+        if (_boostParams.gasUsed != 0) {
+            supplyAmount = _takeFee(
+                GasFeeTakerParams(
+                    _boostParams.gasUsed,
+                    collateralAsset,
+                    exchangedAmount,
+                    MAX_DFS_FEE
+                )
+            );
+        } else {
+            supplyAmount = exchangedAmount;
+        }
+
+        // Supply collateral
+        {
+            int256 vatSupplyAmount = toPositiveInt(
+                convertTo18(_boostParams.joinAddr, supplyAmount)
+            );
+            collateralAsset.approveToken(_boostParams.joinAddr, supplyAmount);
+            IJoin(_boostParams.joinAddr).join(urn, supplyAmount);
+
+            IManager(MCD_MANAGER_ADDR).frob(_boostParams.vaultId, vatSupplyAmount, 0);
+        }
+
+        // if we're using flashloan draw debt and repay fl
+        if (_boostParams.flAddr != address(0)) {
+            uint256 daiAmount = _drawDebt(_boostParams, urn, ilk);
+            DAI_ADDR.withdrawTokens(_boostParams.flAddr, daiAmount);
+        }
+
+        // check if collateral is higher, only if part of strategy
+        if (_boostParams.gasUsed != 0) {
```

Comment:
in which case is it possible for collateral to be less than what we begin with? From my understanding we're supplying once and not withdrawing at any point?

Reply:
maybe check if ratio is lower than before always?

Reply:
This is mainly done as sanity checks, if something goes wrong with the execution, if i check if the ratio is lower that might not mean that the boost went through correctly, might have just withdrawn funds, sold and got stuck on proxy. It seemed better to check if collateral has increased.